perun-contract/src/Perun/Offchain/ChannelTxPool/ChannelTxPool.hs

```diff
@@ -0,0 +1,326 @@
+  return . filterValidStart . partitionByChannelToken $ channelTxs
+  where
+    filterValidStart = filter (\(ChannelTxPool _ ctxs) -> any isValidStart ctxs)
+    isValidStart (ChannelTx _ Nothing (Just _)) = True
+    isValidStart _ = False
+
+-- | resolveTx resolves a TxOutRef to the transactions that created and
+-- potentially spent it.
+resolveTx ::
+  (AsPerunError e, AsContractError e) =>
+  TxOutRef ->
+  Contract w s e [ChainIndexTx]
+resolveTx ref = do
+  let txid = txOutRefId ref
+  createdTx <-
+    txFromTxId txid >>= \case
+      Nothing -> do
+        -- Abort if the chainindex is corrupted and could not retrieve
+        -- the necessary transactions.
+        throwing _SubscriptionError NoTxFromTxIdFetchableErr
+      Just r -> return r
+  -- TODO: Check if the TxOutRef was already spent. This cannot be implemented
+  -- currently because the ChainIndex does not support querying for TXs that
+  -- consumed some UTXO.
+  --
+  -- utxoIsSpent ref >>= \case
+  --    Nothing -> return [createdTx]
+  --    Just createdTx -> return [createdTx, spentTx]
+  return [createdTx]
+
+partitionByChannelToken ::
+  [(ChannelToken, ChannelTx)] ->
+  [ChannelTxPool]
+partitionByChannelToken chanTxsWithToken = map (uncurry ChannelTxPool_) $ Map.toList poolMap
+  where
+    poolMap = foldl' f Map.empty chanTxsWithToken
+    f acc (token, chanTx) = Map.insertWith (++) token [chanTx] acc
+
+mkChannelTx ::
+  (AsPerunError e, AsContractError e) =>
+  ChannelID ->
+  ChainIndexTx ->
+  Contract w s e (Either ChannelTxErr (ChannelToken, ChannelTx))
+mkChannelTx cid citx = do
+  allInputTxs <- concat <$> mapM (resolveTx . txInRef) (citx ^. citxInputs)
+  i <- case resolveInput cid citx allInputTxs of
+    Right i -> return $ Just i
+    Left NoChannelInputErr -> return Nothing
+    Left err -> throwing _SubscriptionError $ ChannelErr err
+  o <- case resolveOutput cid citx of
+    Right o -> return $ Just o
+    Left NoChannelOutputErr -> return Nothing
+    Left err -> throwing _SubscriptionError $ ChannelErr err
+  ct <- retrieveChannelToken i o
+  fmap (ct,) <$> mkChannelTx' i o
+  where
+    retrieveChannelToken ::
+      (AsPerunError e, AsContractError e) =>
+      Maybe (TxOutRef, ChannelAction, ChannelDatum) ->
+      Maybe (TxOutRef, ChannelDatum) ->
+      Contract w s e ChannelToken
+    retrieveChannelToken Nothing Nothing = throwing _SubscriptionError NoChannelTokenAvailableErr
+    retrieveChannelToken (Just (_, _, d)) Nothing = return $ channelToken d
+    retrieveChannelToken Nothing (Just (_, d)) = return $ channelToken d
+    retrieveChannelToken (Just (_, _, d)) (Just (_, d')) = do
+      let iToken = channelToken d
+          oToken = channelToken d'
+      unless (iToken == oToken) $
+        throwing _SubscriptionError MismatchedChannelTokenErr
+      return iToken
+    -- mkChannelTx' returns ChannelTxErr if neither an input, nor an output for
+    -- a channel was found. This happens if the tx in question was randomly
```

Comment:
```suggestion
    -- a channel was found. This happens if the tx in question was maliciously
```

Feel free to deny opr rephrase this if you disagree. I somehow want to emphasize that this is very unlikely to happen "on accident" or "randomly", and is likely an attempt to attack the channel.

Reply:
I like that.